gcc/jit/jit-playback.cc

```diff
@@ -290,6 +290,10 @@ get_tree_node_for_type (enum gcc_jit_types type_)
       return double_type_node;
     case GCC_JIT_TYPE_LONG_DOUBLE:
       return long_double_type_node;
+    case GCC_JIT_TYPE_FLOAT16:
+      return float16_type_node;
+    case GCC_JIT_TYPE_FLOAT128:
+      return float128_type_node;
```

Comment:
According to the [doc](https://gcc.gnu.org/onlinedocs/gcc/Floating-Types.html), those types might not always be supported, so you probably need something like this:
```suggestion
    case GCC_JIT_TYPE_FLOAT16:
      scalar_float_mode mode;
      if (targetm.floatn_mode (16, false).exists (&mode) && GET_MODE_PRECISION (mode) == 16)
        return float16_type_node;
    case GCC_JIT_TYPE_FLOAT128:
      scalar_float_mode mode;
      if (targetm.floatn_mode (128, false).exists (&mode) && GET_MODE_PRECISION (mode) == 128)
        return float128_type_node;
```
I'm not sure if the second part of the checks `GET_MODE_PRECISION (mode) == 128` is required, to be honest.
It seems some platforms have smaller sizes for those mode. We'll need to check this.

Reply:
`float16_type_node` is defined as `global_trees[TI_FLOAT16_TYPE]` and `global_trees` is a global variable, so its values are initialized to NULL. 
In the loop defined here https://github.com/rust-lang/gcc/blob/master/gcc/tree.cc#L9572 the conditions mentioned by you are executed and if they are met, `global_trees[TI_FLOAT16_TYPE]` is initialized to NOT NULL.
So the above `case` statement will return NULL if the value in `global_trees` is NULL. For unsupported type we can add the following condition:
```
if (float16_type_node == NULL) {
  add_error(NULL, "gcc_jit_types value unsupported on this target: %i", type_);
}
```

Reply:
Addressed

Reply:
I'm still not sure if we that could return floating point types that are smaller.
If that's the case, we would still need to add a condition with `GET_MODE_PRECISION (mode) == 16`, for instance.

Reply:
OK, let's add this condition.

Reply:
I'd like to check first if this is indeed necessary. I'm not sure yet.

Reply:
From this code https://github.com/rust-lang/gcc/blob/master/gcc/tree.cc#L9577-L9583

- if the real format precision (`min_precision`) is greater than `GET_MODE_PRECISION(mode)` we take a greater value;
- `gcc_assert` is called when the real format precision (`min_precision`) is not, e.g. 16, 32, 64, 128

So I would say that for _Float16, ..., __float128 floating point types should not be smaller. 
That's my understanding.

Reply:
Ok, that makes sense but it seems they could be bigger if `GET_MODE_PRECISION` returns a bigger precision.
I'm not sure if this is possible, though.

Reply:
From the code, the floating type can't be smaller than declared value, e.g. `16` for `_Float16`. But we are not sure whether `GET_MODE_PRECISION()` can return a value greater than the declared one.
The following condition/safeguard has been added:
```
if (float16_type_node == NULL || TYPE_PRECISION(float16_type_node) != 16)
{
  add_error(...);
  return NULL;
}
return float16_type_node;
```

Reply:
Fixed